### ISSUE TITLE
Upgrade to CouchDB v3

### ISSUE THREAD

**Issue by garethbowen** [medic/cht-core#6289]
CouchDB 3.0 has been released and eventually we'll want to migrate to it.

- Investigate technical compatibility with the CHT and raise issues to fix the bugs, tests, documentation, etc in a backwards compatible way.
- Run performance and scalability tests to determine the extent of any improvements or regressions. Wherever possible raise issues against CouchDB where performance has regressed to give their team plenty of time to investigate before we're ready to upgrade.

This issue can be closed once we have confidence that we have identified and documented all issues that are blocking us from upgrading today.

**Comment by garethbowen** [comment]
I upgraded to Couch 3 locally and the `medic` db became available to all. Fortunately our server checks found it and refused to start API. The problem was that I had no roles configured for the medic database so it was considered public. The workaround was to add all my roles to the `medic` db.

**Comment by dianabarsan** [comment]
I can confirm. 
It seems like the `require_valid_user` doesn't do anything anymore and all database security comes from it's own `_security`. Additionally, when creating new databases, I've noticed that they have a default `_security` that only includes the `_admin` role. 

Additionally, the `_security` object of the `_users` database cannot be changed and only admins can access it. 
Non-admin users can't create new databases.
When an admin creates a new database, it has a default `_security` object that only allows `_admin` role to access the DB (`_admin` is a member and an admin).

**Comment by Ths2-9Y-LqJt6** [comment]
Hey all! I'm a developer looking to possibly help with CHT.  My first step was to see how good your [developer set up docs](https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md) were. In the happy path, they're good! But, being on Ubuntu 18.04, the docs are now a sad path because of CouchDB 3.0 being the default.  Since this ticket is about CouchDB 3.0, commenting here, but happy to open a "improve the dev docs" ticket/PR if that's more helpful.

**tl;dr** - While it looks like this is a CouchDB docs/install issue, maybe we should call out that for Ubuntu >18 you need to be extra careful to install CouchDB 2 like we do for MacOS and homebrew?  I know that we very clearly say `CouchDB 2.x` in our [dependencies](https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md#dependencies), but as non-CouchDB dev who needed to install everything from scratch, this bit me.

**Not so tl;dr** - We correctly deep link to [CouchDB 2.x install docs](http://docs.couchdb.org/en/2.3.1/install/index.html), but then when you click through to [Install on Unix-like systems 1.1.1](https://docs.couchdb.org/en/2.3.1/install/unix.html), you're instructed to run this:

```
echo "deb https://apache.bintray.com/couchdb-deb xenial main" 
    | sudo tee -a /etc/apt/sources.list
curl -L https://couchdb.apache.org/repo/bintray-pubkey.asc \
    | sudo apt-key add -
sudo apt-get update && sudo apt-get install couchdb
```

The issue, I think, is that `apt` goes for the latest the `.deb` for xenial in the [bintray.com](https://apache.bintray.com/couchdb-deb/dists/xenial/main/binary-amd64/) site, which happens to be 3.0 since the [late Feb release](https://blog.couchdb.org/2020/02/26/3-0/). In turn, and the reason I'm commenting on this ticket, is that get an error when I run `grunt dev-api` (pasting full error in case another dev finds it helpful) :

```
  [stack]:                                                                                          
   'StatusCodeError: 403 - {"error":"forbidden","reason":"You can\'t edit the security object of the user database."}\n    at new StatusCodeError (/root/cht-core/api/node_mo
dules/request-promise-core/lib/errors.js:32:15)\n    at Request.plumbing.callback (/root/cht-core/api/node_modules/request-promise-core/lib/plumbing.js:104:33)\n    at Reque
st.RP$callback [as _callback] (/root/cht-core/api/node_modules/request-promise-core/lib/plumbing.js:46:31)\n    at Request.self.callback (/root/cht-core/api/node_modules/req
uest/request.js:185:22)\n    at Request.emit (events.js:198:13)\n    at Request.<anonymous> (/root/cht-core/api/node_modules/request/request.js:1161:10)\n    at Request.emit
 (events.js:198:13)\n    at IncomingMessage.<anonymous> (/root/cht-core/api/node_modules/request/request.js:1083:12)\n    at Object.onceWrapper (events.js:286:20)\n    at In
comingMessage.emit (events.js:203:15)' }                                                        
[nodemon] app crashed - waiting for file changes before starting...  
```

My work around was to [install the snap](https://docs.couchdb.org/en/latest/install/snap.html) which gets me 2.3.1. This will of course break again when the CouchDB snap is updated because:
> Snaps update automatically, and by default, the snapd daemon checks for updates 4 times a day - [snapcraft.io](https://snapcraft.io/docs/keeping-snaps-up-to-date)

so the work around is only buying us so much time ;)

**Comment by garethbowen** [comment]
Thanks for the update! We'll definitely have to fix the code that sets security [as Diana noted above](https://github.com/medic/cht-core/issues/6289#issuecomment-601187399) CouchDB has locked down who can update the _users db (which is a good thing!).

Please raise an issue to update the dev install instructions. We should at least make it clear not to install CouchDB 3.x, but I think we can go one step further by recommending people use a Docker CouchDB container, then we can provide instructions for installing the exact version we support, and it'll be more consistent across OSes.

**Comment by Ths2-9Y-LqJt6** [comment]
cool!  I've submitted #6381 to track the developer doc improvements.

**Comment by dianabarsan** [comment]
This branch contains the so far discovered necessary changes to support CouchDB 3: https://github.com/medic/cht-core/tree/3.12-support-for-couchdb-3 . 

The necessary changes are minimal because of a fix in CouchDB 3.1 (https://github.com/apache/couchdb/issues/2643). Now we can now use a new config option to make the users database security editable, allowing non-admin users to create / delete users, which is a feature that we already support. 
```
[couchdb]
users_db_security_editable = true
```

An additional config change I had to make was to increase the `couchdb.max_document_size` to allow for some larger docs to exist. The default value was too low. 

The changes were used for the CouchDB 3 scalability testing, so, at least, replication works as expected, but I believe no in depth testing was made.

**Comment by dianabarsan** [comment]
After seeing some contradicting data when observing performance differences between CouchDb v2 and Couchdb v3, I ran some more consistent tests. 

Initially I had observed that, locally, on my (quite powerful 24 core CPU) Couch 3 performed radically better,  with performance gains of more than 50% across the board, with some actions being even 10x faster. 
However, when running e2e tests in CI, I would see significant test instability due to operations (seemingly) being slower. 

I wrote a script (details in the [readme](https://github.com/dianabarsan/couchdb-perf)) that spins up couchdb containers (using `medic` images) and measures how long it takes to execute certain targeted queries/endpoints, which are contribute to most of the load of CHT servers. I ran this script over a plethora of tiers of AWS EC2 instances to assess performance in a focused way: 

Results doc: https://docs.google.com/spreadsheets/d/1pUQoIC_jdZxTugLCWFJTppy2LRSygloyof9BRoNrHl4/edit?usp=sharing

![image](https://github.com/medic/cht-core/assets/35681649/a75fb1e3-f1d8-40f8-a264-da40d655d92b)

**Comment by garethbowen** [comment]
Delayed to 4.4.0 to give us more time to test thoroughly and solve the [multi-node upgrade](https://github.com/medic/cht-core/issues/7970) issue.

**Comment by dianabarsan** [comment]
Hi @Reagan 
Did you have a change to give this a try?

**Comment by dianabarsan** [comment]
Merged to `master`. CHT now uses CouchDb v 3.3.2
